Tutorials/WinForms/Classic/implement-crud-functionality-with-xpo-objects.md

```diff
@@ -0,0 +1,169 @@
+                form.ShowDialog(this);
+            }
+        }
+    }
+    ```
+* Press the **F5** key to run the application, double-click a `GridView` record, and see the result.
+* Close both windows and open the `EditCustomerForm` designer.
+* Select `LayoutControl` and add `SimpleButton` from Toolbox. Add `EmptySpaceItem` elements to adjust the button position and size.
+* Rename the `simpleButton1` control to **btnSave** and set its `Text` property to **&Save** (the ampersand sign assigns a mnemonic command to the button).
+* Double click the `btnSave` control to add the `Click` event handler (you can use the **Properties** window to add it).
+* Add this code to the `btnSave_Click` event handler:
+    ```csharp
+    private void btnSave_Click(object sender, EventArgs e) {
+        UnitOfWork.CommitChanges();
+        Close();
+    }
+    ```
+* The [UnitOfwork.CommitChanges](https://docs.devexpress.com/XPO/DevExpress.Xpo.UnitOfWork.CommitChanges) method saves changes to the database. To refresh the CustomersList Form, change code in the `CustomersGridView_RowClick` method as follows: 
+    ```csharp
+    private void Form1_Load(object sender, EventArgs e) {
+        Reload();
+    }
+    private void CustomersGridView_RowClick(object sender, RowClickEventArgs e) {
+        if (e.Clicks == 2) {
+            e.Handled = true;
+            int customerID = (int)CustomersGridView.GetRowCellValue(e.RowHandle, colOid);
+            using (EditCustomerForm form = new EditCustomerForm(customerID)) {
+                form.ShowDialog(this);
+                Reload();
+                CustomersGridView.FocusedRowHandle = CustomersGridView.LocateByValue("Oid", form.CustomerID.Value);
+            }
+        }
+    }
+    private void Reload() {
+        CustomersBindingSource.DataSource = new XPCollection<Customer>(new Session());
+    }
+    ```
+* Use the Visual Studio **Refactor** tool to rename the `Form1_Load` event handler to `CustomersListForm_Load`. To do this, put the cursor at the method name and click the **Edit > Refactor > Rename** menu item or use **Ctrl+R,Ctrl+R**.   
+* Run the application, open the edit Form, change something, and click the **Save** button to see the result.
+## Handle concurrent changes
+>You can skip this part if you are using [InMemoryDataStore](https://docs.devexpress.com/XPO/DevExpress.Xpo.DB.InMemoryDataStore) or an embedded database.
+### Test with multiple users 
+* Build the project and open the output folder (*bin\Debug*) in the file manager. Locate the application executable file and run two application instances.
+* Double click a record in one window to open the edit dialog, and double click the same record in another window.
+* Type something and save changes in each window. The first application saves changes successfully, but the second fails with the following error message: *"Cannot persist the object. It was modified or deleted (purged) by another application"*.
```

Comment:
```suggestion
* Type something and save changes in each window. The first application saves changes successfully, but the second should fail and show the following error message: *"Cannot persist the object. It was modified or deleted."*.
```

Reply:
>"Cannot persist the object. It was modified or deleted (purged) by another application"
This message is from the XPO library. We cannot change it.

Reply:
Ok